Adds an event for rendering blocks that don't exist

Haven-King opened the pull request:
I had some issues getting it to accept non-block `BakedModel`s and loose quads, so I opted for a more general solution. Includes an implementation using Indigo's `TerrainRenderContext`. The event itself needed to be somewhat vague, so that it can be implemented in custom renderers as well.

I'll admit the names aren't great, and would be very open to suggestions. The ability to bake block models has many uses, both mine (large scale world decorations) and for `BlockEntity`s that render other blocks statically to bake them for performance reasons.

Haven-King committed:
Adds an event for rendering blocks that don't exist. Note that they also respect the transformations passed via `context.matrixStack`, so it can be used for some more generic renderings as well.

grondag commented on src/main/java/grondag/frex/api/event/ConditionalEvent.java:
```diff
@@ -0,0 +1,36 @@
+     * @param condition Condition under which the listener should be fired.
+     * @param listener The desired listener.
+     */
+    public abstract void register(Function<C, Boolean> condition, T listener);
```
Why not `Predicate<C>` here instead of `Function<C, Boolean>`?  Would avoid any boxing and is idiomatic.

grondag commented on src/main/java/grondag/frex/mixin/client/render/chunk/ChunkBuilderMixin.java:
```diff
@@ -0,0 +1,55 @@
+        matrixStack.push();
+        matrixStack.translate((blockPos3.getX() & 15), (blockPos3.getY() & 15), (blockPos3.getZ() & 15));
+
+        BakedChunkSectionRenderer.EVENT.invoker().bake(chunkRenderContext);
```
If I'm reading this correctly it gets called for every block position in the section, is that right?

If so, performance-wise that's going to be too heavy.  Would be better to have it call 1X per section and let the even subscriber choose which blocks they want to deal with.  And also need to avoid allocation on each call.

grondag commented:
I made a couple comments and would like your thoughts on those.  I have a better notion now of what you want it to do.

I'd also be curious to know what sort of issues you ran into with loose quads - I assume that was with quad emitter.

Lastly, needs to be clear to event subscribers that they must be thread-safe.  The event itself should be fine in that regard assuming no-one tries to register after init, which they should not do.

Probably more to say but lack the time right now.

Haven-King committed:
- Replaces `Function<C, Boolean>` with `Predicate<C>`.
- Allocates one `ChunkRenderContext` per rebuild task and initializes it with the appropriate variables when needed.

Haven-King committed:
- Fixes skipping condition check if there is only one handler registered to a conditional event.

Haven-King commented:
For context, the following code is used to create the effect below:
```java
BakedChunkSectionRenderer.EVENT.register(context -> {
  int min = Math.min(context.startPos.getY(), context.endPos.getY());
  int max = Math.max(context.startPos.getY(), context.endPos.getY());
  return ((min <= 10 && max >= 10) || (min <= 70) && max >= 70)
      && RegistryUtil.dimensionMatches(MinecraftClient.getInstance().world, ShatteredSky.DIMENSION_TYPE);
}, context -> {
  int min = Math.min(context.getChunkSectionStartPos().getY(), context.getChunkSectionEndPos().getY());
  int max = Math.max(context.getChunkSectionStartPos().getY(), context.getChunkSectionEndPos().getY());

  BlockPos pos = new BlockPos(
      context.getChunkSectionStartPos().getX(),
      (min <= 10 && max >= 10) ? 10 : 70,
      context.getChunkSectionStartPos().getZ());

  BlockState state = ShatteredSky.Blocks.ATMOSPHERE_BLOCK.getDefaultState();

  MatrixStack matrixStack = context.getMatrixStack();
  BlockPos.iterate(pos, pos.add(15, 0, 15)).forEach(blockPos -> {
    matrixStack.push();
    matrixStack.translate(blockPos.getX() & 15, blockPos.getY() & 15, blockPos.getZ() & 15);
    context.getRenderer().bake(blockPos, state);
    matrixStack.pop();
  });
});
```
![image](https://user-images.githubusercontent.com/54509200/102156889-609ba780-3e33-11eb-9994-1f285110d533.png)

grondag commented:
This is looking good with the latest changes.  I want to pull it into FREX for now but it is small enough that Fabric API might actually consider it.

To that end, you can make it smaller still by removing conditional event classes and using the normal Fabric ArrayBackedEvent and baking the predicate into the registration method, like so (sketching, code may not actually compile):

```java
public static void register(Predicate<ChunkRenderContext> condition, BakedChunkSectionRenderer listener) {
    EVENT.register(ctx -> {
        if (condition.test(ctx) {
            listener.bake(ctx);
        }
    }
}
```

I'm also iffy on the naming but will have another closer look when I have a larger block of time and try to get this merged quickly.

Addendum:  if I made the suggested change, I would probably offer two registration methods - one that always fires and one with a predicate, which is basically a convenience method because listeners could always choose to put their condition checks (if any) in the the listener itself.

Addendum to the addendum:  I guess I already did that, because one can simply subscribe to the EVENT directly in the customary manner if a predicate isn't needed.

One more thing: If you want to submit this to FAPI, I'm guessing they will want `ChunkRenderContext` to be an interface with a non-API implementation in a separate package.  That's probably best for FREX, too.  It's at least possible Canvas or Sodium could benefit somehow from rolling their own implementations. (Canvas already has more than one similar context object that could probably be extended to implement a `ChunkRenderContext` interface and avoid an extra init call.)

Haven-King commented:
Regarding removing the conditional event class, one of the main reasons for keeping it is the way FAPI's builtin `ArrayBackedEvent` handles single subscribers, which is to bypass the invoker factory entirely and just invoke the first entry in `handlers`. I could maybe circumvent this by overriding the `invoker` method, but that would still require its own class.

Regarding the addendum and the possibility for unconditional event firing:
Because the condition is used not just for the baking itself, but also for marking empty chunks to be baked (and causing the rebuild task to iterate over the whole chunk), I thought that it would be best for people who _needed_ to render something in every section to be explicit about it.

That being said, I can go ahead and make that move for `ChunkRenderContext` later today.

grondag commented:
> Regarding removing the conditional event class, one of the main reasons for keeping it is the way FAPI's builtin ArrayBackedEvent handles single subscribers, which is to bypass the invoker factory entirely and just invoke the first entry in handlers. I could maybe circumvent this by overriding the invoker method, but that would still require its own class.

I didn't understand this bit.  If the every entry handler includes the predicate check because you wrapped the handler with the predicate check in the registration method, how does the factory (or lack of one) matter?

Haven-King committed:
- Migrates several classes to interfaces so that renderers can implement them independently
- Removes the `ConditionalEvent` class and instead bakes conditions into the callbacks themselves
- Moved some things to the `impl` package

Haven-King committed:
Merge branch '1.16' of https://github.com/grondag/frex into 1.16

Haven-King commented:
I believe I've addressed all of the concerns you had in this commit, with the exception of unconditional event firing, which I still think is not a good idea for performance reasons.
